pkg/helpers/struct.go

```diff
@@ -208,50 +208,87 @@ func AsSlice(x *messages.ListValue) []interface{} {
 // When converting an int64 or uint64 to a NumberValue, numeric precision loss
 // is possible since they are stored as a float64.
 func NewValue(v interface{}) (*messages.Value, error) {
-	switch v := v.(type) {
-	case nil:
+
+	if v == nil {
 		return NewNullValue(), nil
-	case bool:
-		return NewBoolValue(v), nil
-	case int:
-		return NewNumberValue(float64(v)), nil
-	case int32:
-		return NewNumberValue(float64(v)), nil
-	case int64:
-		return NewNumberValue(float64(v)), nil
-	case uint:
-		return NewNumberValue(float64(v)), nil
-	case uint32:
-		return NewNumberValue(float64(v)), nil
-	case uint64:
-		return NewNumberValue(float64(v)), nil
-	case float32:
-		return NewNumberValue(float64(v)), nil
-	case float64:
-		return NewNumberValue(float64(v)), nil
-	case string:
-		if !utf8.ValidString(v) {
+	}
+
+	switch reflect.TypeOf(v).Kind() {
+	case reflect.Bool:
+		return NewBoolValue(v.(bool)), nil
+	case reflect.Int:
+		return NewNumberValue(float64(v.(int))), nil
+	case reflect.Int32:
+		return NewNumberValue(float64(v.(int32))), nil
+	case reflect.Int64:
+		return NewNumberValue(float64(v.(int64))), nil
+	case reflect.Uint:
+		return NewNumberValue(float64(v.(uint))), nil
+	case reflect.Uint32:
+		return NewNumberValue(float64(v.(uint32))), nil
+	case reflect.Uint64:
+		return NewNumberValue(float64(v.(uint64))), nil
+	case reflect.Float32:
+		return NewNumberValue(float64(v.(float32))), nil
+	case reflect.Float64:
+		return NewNumberValue(v.(float64)), nil
+	case reflect.String:
+		refStr := v.(string)
+		if !utf8.ValidString(refStr) {
 			return nil, protoimpl.X.NewError("invalid UTF-8 in string: %q", v)
 		}
-		return NewStringValue(v), nil
-	case time.Time:
-		return NewTimestampValue(v), nil
-	case []byte:
-		s := base64.StdEncoding.EncodeToString(v)
-		return NewStringValue(s), nil
-	case map[string]interface{}:
-		v2, err := NewStruct(v)
+		return NewStringValue(refStr), nil
+	case reflect.Struct:
+		// special case for time values
+		if tVal, ok := v.(time.Time); ok {
+			return NewTimestampValue(tVal), nil
+		}
+		// fallback, attempt to convert struct values
+		interMap := map[string]interface{}{}
+		err := mapstructure.Decode(v, &interMap)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("error decoding struct: %w", err)
 		}
-		return NewStructValue(v2), nil
-	case []interface{}:
-		v2, err := NewList(v)
+		return NewValue(interMap)
+
+	case reflect.Map:
+		reflected := map[string]interface{}{}
+		if mapV, ok := v.(map[string]interface{}); ok {
+			reflected = mapV
+		} else {
+			mapIter := reflect.ValueOf(v).MapRange()
+			for mapIter.Next() {
+				k := mapIter.Key().String() // This will expect a map of key type string; if we get other key types, there will be some weird values here. Not sure if we want to make that a hard error.
```

Comment:
We don't want to silently generate garbage data. Making this a hard error seems preferable so the problem is obvious and not something we have to figure out from garbled data that was ingested.